rustiflow/src/main.rs

```diff
@@ -1102,55 +975,78 @@ fn process_packet_ipv6<T>(
+    features: &BasicFeatures,
+    flow_map: &Arc<DashMap<String, T>>,
+    expirations: &Arc<Mutex<PriorityQueue<String, Instant>>>,
+    lifespan: u64,
+    fwd: bool,
+)
+where
+    T: Flow,
+{
     let flow_id = if fwd {
-        create_flow_id(
-            &source,
-            data.port_source,
-            &destination,
-            data.port_destination,
-            data.protocol,
-        )
+        create_flow_id(&source, source_port, &destination, destination_port, protocol)
     } else {
-        create_flow_id(
-            &destination,
-            data.port_destination,
-            &source,
-            data.port_source,
-            data.protocol,
-        )
+        create_flow_id(&destination, destination_port, &source, source_port, protocol)
     };
 
-    let flow_id_clone = flow_id.clone();
-    let flow_id_remove = flow_id.clone();
-    let mut entry = flow_map.entry(flow_id).or_insert_with(|| {
+    let mut entry = flow_map.entry(flow_id.clone()).or_insert_with(|| {
+
+        expirations.lock().unwrap().push(flow_id.clone(), ts + Duration::from_secs(lifespan));
+
         if fwd {
-            T::new(
-                flow_id_clone,
-                source,
-                data.port_source,
-                destination,
-                data.port_destination,
-                data.protocol,
-                ts_date,
-            )
+            T::new(flow_id.clone(), source, source_port, destination, destination_port, protocol, ts_date)
         } else {
-            T::new(
-                flow_id_clone,
-                destination,
-                data.port_destination,
-                source,
-                data.port_source,
-                data.protocol,
-                ts_date,
-            )
+            T::new(flow_id.clone(), destination, destination_port, source, source_port, protocol, ts_date)
         }
     });
 
+    let flows_to_remove = {
+        let expirations_to_remove = expirations.lock().unwrap();
+        let mut flows_to_remove = Vec::new();
+
+        for (flow_id, &expiration) in expirations_to_remove.iter() {
+            if expiration <= ts {
+                flows_to_remove.push(flow_id.clone());
+            } else {
+                break;
+            }
+        }
+
+        flows_to_remove
+    };
+
+    // Remove the expired flows
+    {
+        let mut expirations = expirations.lock().unwrap();
+
+        for flow_id in flows_to_remove {
+            expirations.remove(&flow_id);
+            flow_map.remove(&flow_id);
+        }
+    }
+
     let end = entry.update_flow(&features, &ts, ts_date, fwd);
     if end.is_some() {
-        export(&end.unwrap());
+        let flow = entry.value();
+        if *NO_CONTAMINANT_FEATURES.lock().unwrap().deref() {
+            export(&flow.dump_without_contamination());
+        } else {
+            export(&flow.dump());
+        }
         drop(entry);
-        flow_map.remove(&flow_id_remove);
+        flow_map.remove(&flow_id);
```

Comment:
```
    if entry.update_flow(&features, &ts, ts_date, fwd).is_some() {
        let flow = entry.value();
        let dump = if *NO_CONTAMINANT_FEATURES.lock().unwrap() {
            flow.dump_without_contamination()
        } else {
            flow.dump()
        };
        export(&dump);
        drop(entry);
        flow_map.remove(&flow_id);
    }

Reply:
Out of the entry.update_flow you already get the output string, what you suggest is just not how the function works.

Reply:
In the code, the output of `update_flow` was also unused. We can reduce code even further:
```
    if let Some(flow) = entry.update_flow(&features, &ts, ts_date, fwd) {
        export(&flow);
        drop(entry);
        flow_map.remove(&flow_id);
    }
```

Reply:
My bad, was indeed incorrect, the update_flow already gives either the contaminant output or the normal one.